Test SpeechSynthesisUtterance volume attribute

guest271314 opened the pull request:
TODO: automate test to verify audio output of speechSynthesis.speak() using Web Audio API.

<!-- Reviewable:start -->

<!-- Reviewable:end -->

Commit:
Test SpeechSynthesisUtterance volume attribute

TODO: automate test to verify audio output of speechSynthesis.speak() using Web Audio API.

w3c-bots commented:
# Build ERRORED

Started: 2018-03-12 02:50:28
Finished: 2018-03-12 02:56:35


<h2>Failing Jobs</h2>
<ul>

<li>lint</li>

<li>chrome:dev</li>

<li>safari:11.0</li>

<li>MicrosoftEdge:16.16299</li>

</ul>




View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/25768)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/352155666)

guest271314 commented:
@andrenatal @fleizach @gshires @jdsmith3000

What needs to be done to add this test to speech-api folder at this repository?

guest271314 commented:
@foolip Can this test be merged into the speech-api tests directory?

foolip commented:
Please have a look at https://web-platform-tests.org/writing-tests/manual.html for some tips for manual tests, in particular the test needs to make it clear to the person running it what is passing and not. For example, you might try to utter the "FAIL" with volume 0, then the word "PASS" with volume 1, and instructions say that the test passes if you *only* hear the PASS.

Commit:
MediaSource crashes tab at Chromium/Chrome

MediaSource crashes tab at Chromium/Chrome
References:
https://github.com/web-platform-tests/wpt/issues/15816
https://bugs.chromium.org/p/chromium/issues/detail?id=820997
https://github.com/guest271314/MediaFragmentRecorder

Commit:
Test HTMLMediaElement.captureStream()

Outputs expected result at Firefox. Crashes tab at Chromium/Chrome https://github.com/web-platform-tests/wpt/issues/15816

Commit:
Merge pull request #2 from guest271314/guest271314-mediasource-htmlmediaelement-capturestream

Test HTMLMediaElement.captureStream()

Commit:
Revert "Test HTMLMediaElement.captureStream()"

Commit:
Merge pull request #3 from guest271314/revert-2-guest271314-mediasource-htmlmediaelement-capturestream

Revert "Test HTMLMediaElement.captureStream()"

foolip commented:
@guest271314 did you accidentally merge in more stuff to this PR? Did you have a look at https://web-platform-tests.org/writing-tests/manual.html? Sorry this has been sitting for so long, but it hasn't come up in my inbox because nothing happened until now.

guest271314 commented:
@foolip It took a few days to get this PR here. Did not intentionally include the Web Speech API `.volume` test, again; as there is already an open PR for that test. Would probably still not be here without the help of @jdm https://github.com/web-platform-tests/wpt/issues/15816#issuecomment-475059112. 

Will try to not to state the issues at GitHub to prevent being banned here as well people tend to get emotional when talking to them without kid gloves on re their "platform"; and have already been redirected to some page that states "abuse" of the site by simply searching, after the recent acquisition of the site. 

Does the test need to be written according to the linked manual test page? And then, does this PR need to be closed and a new PR opened? Can you provide the complete steps necessary for this PR to be incorporated into wpt, to avoid eventually filing another issue for the purpose of navigating this site and `git` process to get to that point?

guest271314 commented:
@foolip Does the manual test need to include buttons for both "PASS" and "FAIL"?

foolip commented:
@guest271314 you don't need to have any buttons in the test itself, just instructions that are visible when opening the page that say what the user should do to judge whether it passes or not. Right now that's in JS comments, so putting the instructions in HTML would be good.

Other than that, you just need to get rid of the extra commits so that there's just one file remaining. If you don't want to deal with `git rebase` and would rather just edit the test in GitHub's UI, I can get rid of the extra file when merging instead.

guest271314 commented:
@foolip Is this appropriate?

```
<!DOCTYPE html>
<html>
<head>
  <title>5.2.3 SpeechSynthesisUtterance volume attribute test - manual</title>
  <script>
var text = "hello universe";
var volumes = [0, 0.16666666666666666, 0.3333333333333333, 0.5, 0.6666666666666666, 0.8333333333333333, 0.9999999999999999];
 window.speechSynthesis.cancel();
 function handleVoicesChanged() {
  volumes.forEach(function(volume) {
    var utterance = new SpeechSynthesisUtterance();
    utterance.text = "hello universe";  
    utterance.volume = volume;
    window.speechSynthesis.speak(utterance);
  });
};
window.speechSynthesis.onvoiceschanged = handleVoicesChanged;
window.speechSynthesis.getVoices();
</script>
</head>
<body>
<p>
Specification

https://w3c.github.io/speech-api/speechapi.html#utterance-attributes

volume attribute

This attribute specifies the speaking volume for the utterance. 
It ranges between 0 and 1 inclusive, with 0 being the lowest volume and 1 the highest volume, with a default of 1. 
If SSML is used, this value will be overridden by prosody tags in the markup.
</p>
<p>
Test

Does the volume of audio output change?
</p>
</body>
</html>
```

If so, what needs to be done? Place that HTML in own fork and make another PR? Or can the code in this PR be changed?

foolip commented:
@guest271314 yes you can update the code in this PR by updating the master branch of your fork, since that's the branch used to create this PR. Or, if you find it simpler, you can just send another PR.

In the code you pasted, using the `voiceschanged` event looks unnecessary, can't the test just start calling `speechSynthesis.speak` right away?

guest271314 commented:
@foolip Has been several OS's ago since dove into Web Speech API. `speak()` communicates with the local Speech Dispatcher server https://github.com/brailcom/speechd. At Chromium if no voices are loaded an empty array could be returned for `getVoices()`, see https://stackoverflow.com/q/46863170. Chrome is shipped with voices created by *oogle. Though the code base is Chromium source, Chromium is not shipped with the custom *oogle voices. Yes, would suggest to use `onvoiceschanged` event. It is not straightforward how the procedure works in the specification.

Have to get the code working again, locally, which may require setting up Speech Dispatcher server again. Been a while since mapped a flow chart of how to use Web Speech API.

Have not, in general, had success filing PR's. Something invariably is included or not included which should or should not have been included. Will try to make sure the code still works at Chromium and update this PR over the next couple days.

guest271314 commented:
@foolip Another consideration that needs to be taken into account is that autoplay policies of browsers have changed since this PR was initially filed, e.g., https://github.com/guest271314/SpeechSynthesisRecorder/issues/13

guest271314 commented:
@foolip Currently `$ spd-conf -u` outputs `bash: /usr/lib/command-not-found: Input/output error` at *nix and no audio output at either Chromium 73 or Nightly 68 and `$ espeak  --stdin 'hello'` outputs
`Empty _dict file: '/usr/lib/i386-linux-gnu/espeak-data/en_dict`. Will have to dive back in to Web Speech API and configuration details at Chromium and *nix.

Commit:
Update WPT Manual Test

guest271314 commented:
@foolip Updated the code at https://github.com/web-platform-tests/wpt/pull/9963/commits/f5803346fdb2d02b77633f9f31273bd99dae3e80. Tested at Firefox 66 and Chromium 73. Will try at Nightly 68 within a day or so. The volume does appear to change at Firefox, not at Chromium plnkr https://plnkr.co/edit/iV2DdU?p=preview.

guest271314 commented:
@foolip Note, for the code to have audio output at Chromium at *nix `espeak`, `espeak-ng` or other TTS program needs to be installed and configured to be used with `speech-dispatcher`, and Chromium needs to be launched with `--enable-speech-dispatcher` flag.

Commit:
Delete MediaSource-MediaStream test

Delete testRecordingMediaFragmentsWithMediaSourceAndMediaRecorder.html from PR including Web Speech API test

Commit:
Update SpeechSynthesisUtterance-volume-attribute-test-manual.html

guest271314 commented:
@foolip Made the suggested changes to the code.

foolip committed:
Rename SpeechSynthesisUtterance-volume-attribute-test-manual.html to SpeechSynthesisUtterance-volume-manual.html

foolip committed:
Update SpeechSynthesisUtterance-volume-manual.html

guest271314 reviewed (commented):
@foolip There is a syntax error at closing `</html>` tag

foolip commented:
@guest271314 I fixed that (and a blank line, and the filename) before merging.

guest271314 commented on mediacapture-fromelement/testRecordingMediaFragmentsWithMediaSourceAndMediaRecorder.html:
```diff
@@ -0,0 +1,187 @@
+<html>
+
+<head>
+  <title>Test recording media fragments using MediaSource and MediaRecorder</title>
```
@foolip Deleted the `MediaSource`/`MediaStream` test.

guest271314 commented on speech-api/SpeechSynthesisUtterance-volume-attribute-test-manual.html:
```diff
@@ -0,0 +1,65 @@
+  </style>
+  <script>
+    const text = "hello universe";
+    const volumes = [0, 0.16666666666666666, 0.3333333333333333, 0.5, 0.6666666666666666, 0.8333333333333333, 0.9999999999999999];
```
@foolip Is the suggestion what is referred to in these reviews as a "nit"? What is the fundamental difference between ` [0, 0.16666666666666666, 0.3333333333333333, 0.5, 0.6666666666666666, 0.8333333333333333, 0.9999999999999999]` and `[0, 0.2, 0.4, 0.6, 1]`? How something "looks" is subjective and not relevant to what something does or does not do. The purpose is to test the entire specified range. Could also test each fraction from 0 to 1 instead of only incrementing by ~1.3. Changed to the suggested range nonetheless.

guest271314 commented:
@foolip How to create a PR for the deleted file https://github.com/guest271314/web-platform-tests/commit/cc69d844995a973785df73ab6157253a239ac8a4 ?

guest271314 commented:
@foolip How to create a PR for the deleted `MediaSource` and `captureStream()` test (one of these branches https://github.com/guest271314/web-platform-tests/tree/guest271314-mediasource-htmlmediaelement-capturestream or https://github.com/guest271314/web-platform-tests/tree/guest271314-mediasource-capturestream which are the same HTML file)?

guest271314 commented:
@foolip `const text = "hello universe"` is not used.

foolip commented:
@guest271314 can you send another PR to fix?

guest271314 commented:
@foolip Still not versed enough yet with how to make a PR using the GitHub GUI. Tried for at least an hour yesterday to create a PR for the deleted (from this PR) `MediaSource`/`HTMLMediaElement.captureStream()` test. Every time the displayed output was, paraphrasing, "nothing to compare".  Will continue to try.

guest271314 commented:
@foolip Following the suggestion of @jdm at https://github.com/web-platform-tests/wpt/issues/15816#issuecomment-475059112 was helpful though led to multiple PR's being incorporated into this PR (which had to delete files from). https://github.com/web-platform-tests/wpt/compare/master...guest271314:master is the link from that comment. How to "clean up" the multiple PR sources to both update this PR (the code) and separately file the PR which tests capturing `<video>` using `.captureStream()` where `src` attribute is set to `MediaSource` instance?

foolip commented:
@guest271314, if you want to edit a single file and create a new PR, navigate to it from https://github.com/web-platform-tests/wpt?files=1 and then edit, you'll be at a URL like this:
https://github.com/web-platform-tests/wpt/edit/master/speech-api/SpeechSynthesisUtterance-volume-manual.html

Then just write a commit message and create the PR.

guest271314 commented:
@foolip Filed updated PR. How to file the PR for the `MediaSource`/`HTMLMediaElement.captureStream()` test https://github.com/web-platform-tests/wpt/pull/9963/commits/cc69d844995a973785df73ab6157253a239ac8a4 ?

guest271314 commented:
@foolip  These are the two commits that am trying to file a PR for each https://github.com/web-platform-tests/wpt/commit/4ec675380896ad787270e4a46430d691fbb6f786 and https://github.com/web-platform-tests/wpt/commit/cc69d844995a973785df73ab6157253a239ac8a4. The former is for the mediacapture-fromelement directory the latter is for media-source directory.

guest271314 commented:
@foolip This is the filed PR https://github.com/web-platform-tests/wpt/pull/16756. Probably needs some cleaning up, though not sure how to do so
